Shared Axis Transition

shihaohong opened the pull request:
## Description
Implement the shared x-axis and shared y-axis transitions. This PR also deletes the existing `SharedZAxisTransitionsBuilder` and pulls it into `SharedAxisPageTransitionsBuilder` through an option called `SharedAxisTransitionType.scaled`.

```dart
MaterialApp(
  theme: ThemeData(
    pageTransitionsTheme: PageTransitionsTheme(
      builders: {
        TargetPlatform.android: SharedAxisPageTransitionsBuilder(
          transitionType: SharedAxisTransitionType.horizontal, // or SharedAxisTransitionType.vertical
        ),
      },
    ),
  ),
  // ...
);
```

### Shared X-Axis Transition
![Shared X-axis Transition](https://user-images.githubusercontent.com/27032613/71688757-5634ea80-2d55-11ea-8699-105762ca7b6b.gif)

### Shared Y-Axis Transition
![Shared Y-axis Transition](https://user-images.githubusercontent.com/27032613/71691659-cf840b80-2d5c-11ea-83a8-5773f3e1c9d7.gif)

## Discussion
### Similarity to Shared Z Axis Transition
I noticed that the curves used for this transition is identical to https://github.com/flutter/packages/pull/57 minus the difference in scale/translation animations, but decided to create a separate builder and suite of tests in the event that the specs change, which might make it difficult to decouple their implementations. If using the same builder and adding a new enum value of something like `SharedAxisTransition.scaled` seems more convenient, we could implement it that way as well.

## Sample App:
```dart
import 'package:flutter/material.dart';
import 'package:animations/animations.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _selectedIndex = 0;

  final List<Color> _colors = [Colors.white, Colors.red, Colors.yellow];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Page Transition Example'),
      ),
      body: PageTransitionSwitcher(
        transitionBuilder: (
          Widget child,
          Animation<double> primaryAnimation,
          Animation<double> secondaryAnimation,
        ) {
          return SharedAxisTransition(
            animation: primaryAnimation,
            secondaryAnimation: secondaryAnimation,
            child: child,
            transitionType: SharedAxisTransitionType.vertical,
          );
        },
        child: Container(
          key: ValueKey<int>(_selectedIndex),
          color: _colors[_selectedIndex],
          child: Center(
            child: FlutterLogo(size: 300),
          )
        ),
      ),
      bottomNavigationBar: BottomNavigationBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(
            icon: Icon(Icons.home),
            title: Text('White'),
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.business),
            title: Text('Red'),
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.school),
            title: Text('Yellow'),
          ),
        ],
        currentIndex: _selectedIndex,
        onTap: (int index) {
          setState(() {
            _selectedIndex = index;
          });
        },
      ),
    );
  }
}
```

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -0,0 +1,436 @@
+
+import 'utils/curves.dart';
+
+/// Determines which type shared axis transition.
```
nit: this sentence seems to be incomplete. Maybe missing an "of"?

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
````diff
@@ -0,0 +1,436 @@
+/// This transition can also be used directly in a
+/// [PageTransitionSwitcher.transitionBuilder] to transition
+/// from one widget to another as seen in the following example:
+/// ```dart
````
nit: add a blank line before this

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -0,0 +1,436 @@
+  ///    property when the it is used as a page transition.
+  final Animation<double> secondaryAnimation;
+
+  /// Determines which type shared axis transition.
```
nit: incomplete sentence?

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -0,0 +1,436 @@
+      BuildContext context,
+      Animation<double> animation,
+      Animation<double> secondaryAnimation,
+      Widget child) {
```
nit: add a trailing comma and reformat.

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -0,0 +1,436 @@
+  @override
+  Widget build(BuildContext context) {
+    final Animatable<Offset> slideInTransition = Tween<Offset>(
+      begin: transitionType == SharedAxisTransitionType.horizontal
```
enums are best used with switch/case

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -0,0 +1,436 @@
+  Widget build(BuildContext context) {
+    final Animatable<Offset> slideOutTransition = Tween<Offset>(
+      begin: Offset.zero,
+      end: transitionType == SharedAxisTransitionType.horizontal
```
same

goderbauer commented on packages/animations/test/shared_axis_transition_test.dart:
```diff
@@ -0,0 +1,1077 @@
+              tester,
+              SharedAxisTransitionType.horizontal,
+            ),
+            0.0);
```
uber-nit (feel free to ignore): I think that dartfmt formats these multi-line expects nicer if you add a trailing comma.

(here and elsewhere)

shihaohong commented:
@goderbauer I addressed the code review feedback. 

We also spoke offline about having the SharedZAxisTransition builder simply be a part of the x and y SharedAxisTransition builders. I've deleted the SharedZAxis transition code and incorporated it into the SharedAxisTransition builder itself.

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -9,18 +9,27 @@ import 'package:flutter/widgets.dart';
-/// to create a page transition with [SharedZAxisTransition].
+/// Determines which type of shared axis transition is used.
+enum SharedAxisTransitionType {
+  /// Creates a shared axis vertical translation page transition.
```
Can you mention the alternative name (x, y, z axis) in these doc strings?

goderbauer commented on packages/animations/lib/src/shared_axis_transition.dart:
```diff
@@ -48,11 +61,11 @@ import 'utils/curves.dart';
 ///         color: Colors.blue,
 ///         child: Center(
-///           child: RaisedButton(
+///           child: MaterialButton(
```
I believe MaterialButton is to be avoided and RaisedButton, FlatButton, or OutlineButton should be used.

goderbauer commented on packages/animations/pubspec.yaml:
```diff
@@ -9,6 +9,7 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
+  vector_math: ^2.0.8
```
This is only used in tests, right? In that case, it should be a dev_dependency and moved to the section below.
